litestar/response/base.py

```diff
@@ -24,28 +22,172 @@
         ResponseHeaders,
         Scope,
         Send,
+        Serializer,
         TypeEncodersMap,
     )
 
+__all__ = ("ASGIResponse", "Response")
+
 T = TypeVar("T")
 
 
+class ASGIResponse:
+    """A low-level ASGI response class."""
+
+    __slots__ = (
+        "background",
+        "body",
+        "content_length",
+        "encoded_headers",
+        "encoding",
+        "is_head_response",
+        "status_code",
+    )
+
+    def __init__(
+        self,
+        *,
+        background: BackgroundTask | BackgroundTasks | None = None,
+        body: bytes = b"",
+        content_length: int | None = None,
+        cookies: list[Cookie] | None = None,
+        encoded_headers: list[tuple[bytes, bytes]] | None = None,
+        encoding: str = "utf-8",
+        headers: dict[str, Any] | None = None,
+        is_head_response: bool = False,
+        media_type: str | None = None,
+        status_code: int | None = None,
+    ) -> None:
+        """A low-level ASGI response class.
+
+        Args:
+            background: A background task or a list of background tasks to be executed after the response is sent.
+            body: encoded content to send in the response body.
+            content_length: The response content length.
+            cookies: The response cookies.
+            encoded_headers: The response headers.
+            encoding: The response encoding.
+            headers: The response headers.
+            is_head_response: A boolean indicating if the response is a HEAD response.
+            media_type: The response media type.
+            status_code: The response status code.
+        """
+        status_code = status_code or HTTP_200_OK
+        cookies = cookies or []
+        encoded_headers = encoded_headers or []
+        headers = headers or {}
+        media_type = media_type or get_enum_string_value(MediaType.JSON)
+
+        status_allows_body = not (
+            status_code in {HTTP_204_NO_CONTENT, HTTP_304_NOT_MODIFIED} or status_code < HTTP_200_OK
+        )
+
+        if not status_allows_body or is_head_response:
+            if body and body != b"null":
+                raise ImproperlyConfiguredException(
+                    "response content is not supported for HEAD responses and responses with a status code "
+                    "that does not allow content (304, 204, < 200)"
+                )
+            body = b""
+
+        encoded_headers.append(
+            (
+                b"content-type",
+                (f"{media_type}; charset={encoding}" if media_type.startswith("text/") else media_type).encode(
+                    "latin-1"
+                ),
+            ),
+        )
+
+        if content_length is None:
+            content_length = len(body)
+
+        if "content-length" not in headers and content_length:
```

Comment:
What if it's in `encoded_headers`? See comment above about those.

Reply:
This problem isn't unique to this PR though.

In main, we do care to not duplicate or overwrite the value using `setdefault()`:

https://github.com/litestar-org/litestar/blob/3c6480e5dca75e7df37247c548ac1fbb1425d717/litestar/response/base.py#L110-L113

So if `"content_type"` is in `headers` on instantiation of the `Response`, that is OK. But generally, headers are not passed to the response as a mapping, rather they are pre-encoded and we don't make any effort to record whether it is set on the layers or not in that case:

https://github.com/litestar-org/litestar/blob/3c6480e5dca75e7df37247c548ac1fbb1425d717/litestar/handlers/http_handlers/_utils.py#L67-L71

Nor any attempt to manage duplicate headers when they are merged with the response headers/cookies:

https://github.com/litestar-org/litestar/blob/3c6480e5dca75e7df37247c548ac1fbb1425d717/litestar/response/base.py#L269-L275

So we are paying a "correctness" price for the pre-encoding optimization. There might be ways to work around this however, without losing the pre-encoding optimization. E.g., we could keep a mapping of pre-encoded header name, to its index in the encoded headers collection that is passed to the asgi response... but IDK.. what if we make `encoded_headers` accept an object that keeps such state?